### ISSUE TITLE
Numerous extraction problems

### ISSUE THREAD

**Issue by Zimmi48** [Zimmi48/bugzilla-test#1460]
Note: the issue was created automatically with bugzilla2github tool

# Bugzilla Bug ID: 820
Date: 2004-07-23 10:09:30 +0200
From: jevgenijs@dva.lv
To:   

Last updated: 2008-10-21 11:11:50 +0200

## Bugzilla Comment ID: 57
Date: 2004-07-23 10:09:30 +0200
From: jevgenijs@dva.lv

Full_Name: Jevgenijs Sallinens
Version: 8.0
OS: Windows XP
Submission from: 81.198.180.1 (81.198.180.1)

On my web site www.mathcoq.dva.lv (see Downloads/ Nat.zip)
there is development for natural numbers and instructions 
for extractions. I have found here numerous problems with extraction:
a) file main_Nat.ml- p1,q1 not required to be extracted since
not used in coq_Nat (these are used to describe properties of eq_nat) 

let coq_Nat =
  let p1 = fun x y z ->
    B_Props.BoolProps.coq_OR (B_Props.BoolProps.coq_NOT (eq_nat x y))
      (B_Props.BoolProps.coq_OR (B_Props.BoolProps.coq_NOT (eq_nat y z))
        (eq_nat x z))
  in
  let q1 = fun x y ->
    B_Props.BoolProps.coq_OR (B_Props.BoolProps.coq_NOT (eq_nat x y))
      (eq_nat y x)
  in
  (fun x x0 -> eq_nat x x0)

b) Modules containing only propositions (n_Props2.ml, Gen_Induction etc.)
should not be extracted since not used in the code.

c) I was forced to duplicate modules like  
Module MBool : I_MBool:=XBool.
in IndBool.v on order to get correct extracted code for MBool.

d) I was forced to change notations SA,SB to RSA,RSB in I_Recursion.v and
related places :
Variable  RSA:Set.
Variable  A:DSetoid RSA.
Variable  RSB:Set.
Variable  B:DSetoid RSB.
Otherwise extraction of N_Props3 and NatProps were producing error
on cyclic defintions like: type coq_SA = coq_SA. 

e) It would be nice to have shorter extractions for description of module
parameters with Definitions (see recursion.ml and many others).

Many thanks,
Jevgenijs Sallinens.
  

## Bugzilla Comment ID: 59
Date: 2004-10-04 18:13:50 +0200
From: jevgenijs@dva.lv

Pierre Letouzey <coq-bugs@pauillac.inria.fr>
Dear Jevgenijs

First I apologize deeply for the abnormal answer delay, due to holidays + 
the move from Paris to Munich for the start of my postdoc. 

Then, thanks for your bug reports: 

a) Solved in Coq CVS. The extraction was optimizing the extracted code 
in a somewhat wrong order (it was deciding to keep or not a let-in *before* 
having optimized the body of this let-in). I can send you a patch if you 
prefer not to look at Coq CVS. 

b) As pointed out by Roland Zumkeller in Bug#843, even an empty module can
be used, for example as argument of a functor. I can probably find a workaround

for that, for example creating an empty module only when needed, but that's
some
extra work. Since the gain would be purely cosmetic, I probably won't do it for

the moment. 

c) I don't really understand the problem. Can you please send me the precise 
instructions you used for extracting, as well as a sample of incorrect
extracted
code for MBool. The Nat.zip I've downloaded only contains some .v files. 

I've tried to reproduce your problem:

Require Import IndBool.
Extraction "test1.ml" XBool.
Extraction "test2.ml" MBool.

But then the two files only differ by the added line for MBool:
$ diff test1.ml test2.ml
235,236d234
< module MBool = XBool
< 

And both files are accepted by ocaml. 

d) Generaly speaking, some parts of module extraction are still experimental, 
and the part concerning naming/renaming/qualifying_names certainly is. 
Moreover I'm quite certain there are no perfect & simple solution: I'm at least
aware of one situation where Ocaml module design is strictly less powerful than

the one of Coq, and forbids a simple translation from Coq names to Ocaml names:

Module M.
 Definition t:=0.
 Module N.
  Definition t:=1.
  Definition u:=M.t
 End N.
End M.

In ocaml, the internal M.t isn't allowed, since M isn't defined yet at that
point. 
But you cannot either use "let u = t" because of the name clash. 

Unfortunately, the problem you encountered seems to be a variant of the
previous:

Module M.
 Definition t:=nat.
 Module N.
  Definition t:=M.t.
 End N.
End M.

No M.t allowed in Ocaml, and no "type t = t" either. 

I currently see no simple general fix to these naming problems (in fact, I do
know 
a complex solution, but haven't the time nor the courage to implement it). So
your 
manual renaming in Coq is currently the best option. One last remark: in your
case
the problem occurs only in the .mli files. You can remove them and regenerate
them
via "ocamlc -c -i". By the way, the solution of ocamlc to the above problem is
to
expand the first "t", and generate "type t = nat" instead of "type t = t".
That's 
ok for the problem about types, but can't be a solution for the initial problem

about terms, where duplication could be really bad. 

e) I cannot for the moment take advantage of Coq syntax: 
                Module Foo : Bar with Definition Zoo := Zii. 
In fact, this syntax is currently only syntactic sugar, available only at user 
level, whereas Coq internal kernel completely unfolds "Bar with Definition ..."
into a complete new module type. And the extraction works by considering
internal 
kernel objects. Making these " ... with Definition ... " visible to Coq kernel
is 
out of my power (Jacek ?), while retreiving them at extraction time is
unfeasible. 

Anyway, I will keep that request in one corner of my mind, but with the mention

"cosmetic feature wish", hence with low priority. 

## Bugzilla Comment ID: 61
Date: 2004-10-05 13:47:44 +0200
From: jevgenijs@dva.lv

jevgenijs@dva.lv
This message is in MIME format.

---MOQ109698406400ce9e33b23825a39994218c65a61a7e
Content-Type: text/plain; charset=ISO-8859-1
Content-Transfer-Encoding: 8bit

Dear Pierre,

Many thanks for your so detailed answer and the work you have done to check
these problems. I have changed my approach, by reducing significantly the usage 
of modules, so probably most of encountered problems will not appear with this 
new solution and could wait for new Coq release. (I am using Windows version 
and since seems there is no available instructions how to compile Coq source 
for Windows, I am just using ready compiled versions). After 2-3 weeks I am 
planing to test extraction of the whole development again. 
Sorry, for probably there was incomplete compile.txt file, at the moment I have
placed it for the first time. Now this file contains instructions (and is 
attached here). 
In short I have used just following:
Require Import NatProps.
Recursive Extraction Library NatProps.

Until the end of this week I will try to repeat the problem (c) and will sent 
you extracted code. Without adding Module XBool there were some problems.

Thank you a lot.

Best Regards,
Jevgenijs.

 

> 
> c) I don't really understand the problem. Can you please send me the precise
> 
> instructions you used for extracting, as well as a sample of incorrect
> extracted
> code for MBool. The Nat.zip I've downloaded only contains some .v files. 
> 
> I've tried to reproduce your problem:
> 
> Require Import IndBool.
> Extraction "test1.ml" XBool.
> Extraction "test2.ml" MBool.
> 
> But then the two files only differ by the added line for MBool:
> $ diff test1.ml test2.ml
> 235,236d234
> < module MBool = XBool
> < 
> 
> And both files are accepted by ocaml. 
> 

-------------------------------------------------
This mail sent through IMP: http://horde.org/imp/

---MOQ109698406400ce9e33b23825a39994218c65a61a7e
Content-Type: text/plain; name="Compile.txt"
Content-Transfer-Encoding: base64
Content-Disposition: attachment; filename="Compile.txt"

Rm9sbG93aW5nIGluc3RydWN0aW9ucyBhcmUgZm9yIE1TIFdpbmRvd3MgdXNlcnMNCihDb3E4LjAg
YW5kIE9iamVjdGl2ZSBDYW1sIGFyZSBhc3N1bWVkIHRvIGJlIHByb3Blcmx5IGluc3RhbGxlZCku
DQogDQoNCkEpIENvbXBpbGUgKi52IGZpbGVzIHdpdGggY29xYy5leGUNCg0KY29xYyAgSV9JbmRC
b29sDQpjb3FjICBJbmRCb29sDQpjb3FjICBTZXRvaWRzDQpjb3FjICBCX1Byb3BzDQpjb3FjICBN
YWluX05hdA0KY29xYyAgTWFwcw0KY29xYyAgQ2FzZXNQcm9kDQpjb3FjICBJX05hdA0KY29xYyAg
SV9JbmR1Y3Rpb24NCmNvcWMgIElfUmVjdXIwDQpjb3FjICBJX1JlY3VyMDINCmNvcWMgIElfUmVj
dXIwMw0KY29xYyAgSV9SZWN1cg0KY29xYyAgSV9SZWN1cjINCmNvcWMgIElfUmVjDQpjb3FjICBJ
X1JlY05hdA0KY29xYyAgSV9SZWNOYXQyDQpjb3FjICBJX1JlY05hdDMNCmNvcWMgIElfUmVjMg0K
Y29xYyAgSV9SZWNOYXRzDQpjb3FjICBJX05fUHJvcHMNCmNvcWMgIElfTl9Qcm9wczINCmNvcWMg
IElfUmVjMV9Qcm9wcw0KY29xYyAgSV9SZWMxX1Byb3BzMg0KY29xYyAgSV9JdGVyYXRpb25zDQpj
b3FjICBJX0l0ZXJQcm9wcw0KY29xYyAgSV9HZW5JbmR1Y3Rpb24NCmNvcWMgIElfUmVjdXJzaW9u
DQpjb3FjICBJX1JlYzMNCmNvcWMgIElfTl9Qcm9wczMNCmNvcWMgIEluZHVjdGlvbg0KY29xYyAg
UmVjdXIwDQpjb3FjICBSZWN1cjAyDQpjb3FjICBSZWN1cjAzDQpjb3FjICBSZWN1cg0KY29xYyAg
UmVjdXIyDQpjb3FjICBSZWMNCmNvcWMgIFJlYzINCmNvcWMgIFJlY05hdHMNCmNvcWMgIE5fUHJv
cHMNCmNvcWMgIE5fUHJvcHMyDQpjb3FjICBSZWMxX1Byb3BzDQpjb3FjICBSZWMxX1Byb3BzMg0K
Y29xYyAgSXRlcmF0aW9ucw0KY29xYyAgSXRlclByb3BzDQpjb3FjICBHZW5JbmR1Y3Rpb24NCmNv
cWMgIFJlY3Vyc2lvbg0KY29xYyAgUmVjTmF0UHJvZA0KY29xYyAgUmVjMw0KY29xYyAgTl9Qcm9w
czMNCmNvcWMgIE5hdFByb3BzDQoNCg0KQikgRXh0cmFjdCBPY2FtbCBmaWxlcyB3aXRoIGNvcXRv
cC5leGUNCg0KUmVxdWlyZSBJbXBvcnQgTmF0UHJvcHMuDQpSZWN1cnNpdmUgRXh0cmFjdGlvbiBM
aWJyYXJ5IE5hdFByb3BzLg0KDQpDKSBDb21waWxlIE9jYW1sIGZpbGVzIHdpdGggb2NhbWxjLmV4
ZQ0KDQpvY2FtbGMgLWMgICBJX0luZEJvb2wubWxpDQpvY2FtbGMgLWMgICBJX0luZEJvb2wubWwN
Cm9jYW1sYyAtYyAgIEluZEJvb2wubWxpDQpvY2FtbGMgLWMgICBJbmRCb29sLm1sDQpvY2FtbGMg
LWMgICBTZXRvaWRzLm1saQ0Kb2NhbWxjIC1jICAgU2V0b2lkcy5tbA0Kb2NhbWxjIC1jICAgQl9Q
cm9wcy5tbGkNCm9jYW1sYyAtYyAgIEJfUHJvcHMubWwNCm9jYW1sYyAtYyAgIE1haW5fTmF0Lm1s
aQ0Kb2NhbWxjIC1jICAgTWFpbl9OYXQubWwNCm9jYW1sYyAtYyAgIE1hcHMubWxpDQpvY2FtbGMg
LWMgICBNYXBzLm1sDQpvY2FtbGMgLWMgICBDYXNlc1Byb2QubWxpDQpvY2FtbGMgLWMgICBDYXNl
c1Byb2QubWwNCm9jYW1sYyAtYyAgIElfTmF0Lm1saQ0Kb2NhbWxjIC1jICAgSV9OYXQubWwNCm9j
YW1sYyAtYyAgIEluZHVjdGlvbi5tbGkNCm9jYW1sYyAtYyAgIEluZHVjdGlvbi5tbA0Kb2NhbWxj
IC1jICAgUmVjdXIwLm1saQ0Kb2NhbWxjIC1jICAgUmVjdXIwLm1sDQpvY2FtbGMgLWMgICBSZWN1
cjAyLm1saQ0Kb2NhbWxjIC1jICAgUmVjdXIwMi5tbA0Kb2NhbWxjIC1jICAgUmVjdXIwMy5tbGkN
Cm9jYW1sYyAtYyAgIFJlY3VyMDMubWwNCm9jYW1sYyAtYyAgIFJlY3VyLm1saQ0Kb2NhbWxjIC1j
ICAgUmVjdXIubWwNCm9jYW1sYyAtYyAgIFJlY3VyMi5tbGkNCm9jYW1sYyAtYyAgIFJlY3VyMi5t
bA0Kb2NhbWxjIC1jICAgUmVjLm1saQ0Kb2NhbWxjIC1jICAgUmVjLm1sDQpvY2FtbGMgLWMgICBS
ZWMyLm1saQ0Kb2NhbWxjIC1jICAgUmVjMi5tbA0Kb2NhbWxjIC1jICAgUmVjTmF0cy5tbGkNCm9j
YW1sYyAtYyAgIFJlY05hdHMubWwNCm9jYW1sYyAtYyAgIE5fUHJvcHMubWxpDQpvY2FtbGMgLWMg
ICBOX1Byb3BzLm1sDQpvY2FtbGMgLWMgICBOX1Byb3BzMi5tbGkNCm9jYW1sYyAtYyAgIE5fUHJv
cHMyLm1sDQpvY2FtbGMgLWMgICBSZWMxX1Byb3BzLm1saQ0Kb2NhbWxjIC1jICAgUmVjMV9Qcm9w
cy5tbA0Kb2NhbWxjIC1jICAgUmVjMV9Qcm9wczIubWxpDQpvY2FtbGMgLWMgICBSZWMxX1Byb3Bz
Mi5tbA0Kb2NhbWxjIC1jICAgSXRlcmF0aW9ucy5tbGkNCm9jYW1sYyAtYyAgIEl0ZXJhdGlvbnMu
bWwNCm9jYW1sYyAtYyAgIEl0ZXJQcm9wcy5tbGkNCm9jYW1sYyAtYyAgIEl0ZXJQcm9wcy5tbA0K
b2NhbWxjIC1jICAgR2VuSW5kdWN0aW9uLm1saQ0Kb2NhbWxjIC1jICAgR2VuSW5kdWN0aW9uLm1s
DQpvY2FtbGMgLWMgICBSZWN1cnNpb24ubWxpDQpvY2FtbGMgLWMgICBSZWN1cnNpb24ubWwNCm9j
YW1sYyAtYyAgIFJlY05hdFByb2QubWxpDQpvY2FtbGMgLWMgICBSZWNOYXRQcm9kLm1sDQpvY2Ft
bGMgLWMgICBSZWMzLm1saQ0Kb2NhbWxjIC1jICAgUmVjMy5tbA0Kb2NhbWxjIC1jICAgTl9Qcm9w
czMubWxpDQpvY2FtbGMgLWMgICBOX1Byb3BzMy5tbA0Kb2NhbWxjIC1jICAgTmF0UHJvcHMubWxp
DQpvY2FtbGMgLWMgICBOYXRQcm9wcy5tbA0Kb2NhbWxjIC1hIEluZEJvb2wuY21vIFNldG9pZHMu
Y21vIEJfUHJvcHMuY21vIE1haW5fTmF0LmNtbyBNYXBzLmNtbyBDYXNlc1Byb2QuY21vIEluZHVj
dGlvbi5jbW8gUmVjdXIwLmNtbyBSZWN1cjAyLmNtbyBSZWN1cjAzLmNtbyBSZWN1ci5jbW8gUmVj
dXIyLmNtbyBSZWMuY21vIFJlYzIuY21vIFJlY05hdHMuY21vIE5fUHJvcHMuY21vIE5fUHJvcHMy
LmNtbyBSZWMxX1Byb3BzLmNtbyBSZWMxX1Byb3BzMi5jbW8gSXRlcmF0aW9ucy5jbW8gSXRlclBy
b3BzLmNtbyBHZW5JbmR1Y3Rpb24uY21vIFJlY3Vyc2lvbi5jbW8gUmVjTmF0UHJvZC5jbW8gUmVj
My5jbW8gTl9Qcm9wczMuY21vIE5hdFByb3BzLmNtbyAgLW8gQ29xTmF0LmNtYQ0KDQpEKSBUZXN0
IGxpYmFyYXJ5IENvcU5hdC5jbWEgd2l0aCBvY2FtbC5leGUNCg0KI3VzZSAiUHJpbnRmLm1sIjs7
DQojbG9hZCAiQ29xTmF0LmNtYSI7Ow0KDQoNCmxldCBvbmUgPSBOYXRQcm9wcy5OQS5jb3FfTl9T
IE5hdFByb3BzLk5BLmNvcV9OX087Ow0KbGV0IHR3byA9IE5hdFByb3BzLk1fTl9Qcm9wcy5jb3Ff
Tl9wbHVzIG9uZSBvbmU7Ow0KbGV0IHRocmVlID0gTmF0UHJvcHMuTV9OX1Byb3BzLmNvcV9OX3Bs
dXMgdHdvIG9uZTs7DQpsZXQgZm91ciA9IE5hdFByb3BzLk1fTl9Qcm9wcy5jb3FfTl9wbHVzIG9u
ZSB0aHJlZTs7DQpwcmludGYgIiViIiAoU2V0b2lkcy5jb3FfRXF1YWwgTmF0UHJvcHMuTkEuY29x
X05hdCAoTmF0UHJvcHMuTV9OX1Byb3BzLmNvcV9OX3BsdXMgdHdvIHR3bykgZm91cik7Ow0KDQoN
Cg0K

---MOQ109698406400ce9e33b23825a39994218c65a61a7e--

## Bugzilla Comment ID: 60
Date: 2004-10-07 21:40:28 +0200
From: jevgenijs@dva.lv

"Jevgenijs Sallinens" <jevgenijs@dva.lv>
This is a multi-part message in MIME format.

------=_NextPart_000_0008_01C4ACB6.43257490
Content-Type: text/plain;
	charset="iso-8859-4"
Content-Transfer-Encoding: quoted-printable

Dear Pierre,

Here are some extracts concerning problem (c).
These are in case when in IndBool.v  one defines MBool directly =20
Module MBool :I_MBool.
End MBool.

Problems dissapear if we let=20
=20
Module XBool :I_MBool.
End XBool.
Module MBool:I_MBool :=3DXBool.

For full extracted files see http://www.mathcoq.dva.lv/downloads/Nat.zip

Sorry if all this just for lack of my understanding of Ocaml compiling.=20

Best Regards,
Jevgenijs.

**************************************
file indBool.mli
**************************************
module MBool :=20
 I_IndBool.I_MBool

**************************************
file indBool.ml
**************************************

module MBool =3D=20
 struct=20
  type ind_bool =3D
    | Coq_ind_B_false
    | Coq_ind_B_true
......

end

**************************************
file B_Props.ml
**************************************

open IndBool
open Setoids
module XBoolProps =3D=20
 struct=20
  module BoolCases =3D=20
   struct=20
......
  let coq_NOT =3D function
    | IndBool.MBool.Coq_ind_B_false -> coq_true
    | IndBool.MBool.Coq_ind_B_true -> coq_false
........
   end
end

**************************************
ocamlc -c   B_Props.ml
File "B_Props.ml", line 70, characters 10-39:
Unbound constructor IndBool.MBool.Coq_ind_B_false
**************************************

**************************************
file Main_Nat.ml
**************************************
  let rec coq_rec f a =3D function
    | IndBool.MBool.Coq_ind_O -> a
    | IndBool.MBool.Coq_ind_S m -> f (coq_rec f a m)
 end
**************************************
ocamlc -c   Main_Nat.ml
File "Main_Nat.ml", line 62, characters 6-29:
Unbound constructor IndBool.MBool.Coq_ind_O

------=_NextPart_000_0008_01C4ACB6.43257490
Content-Type: text/html;
	charset="iso-8859-4"
Content-Transfer-Encoding: quoted-printable

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<HTML><HEAD>
<META http-equiv=3DContent-Type content=3D"text/html; =
charset=3Diso-8859-4">
<META content=3D"MSHTML 6.00.2737.800" name=3DGENERATOR>
<STYLE></STYLE>
</HEAD>
<BODY bgColor=3D#ffffff>
<DIV><FONT face=3DArial size=3D2>Dear Pierre,</FONT></DIV>
<DIV><FONT face=3DArial size=3D2></FONT>&nbsp;</DIV>
<DIV><FONT face=3DArial size=3D2>Here are some extracts concerning =
problem=20
(c).</FONT></DIV>
<DIV><FONT face=3DArial size=3D2>These are in case when in =
IndBool.v&nbsp; one=20
defines MBool&nbsp;directly &nbsp;</FONT></DIV>
<DIV><FONT face=3DArial size=3D2>Module MBool :I_MBool.</FONT></DIV>
<DIV><FONT face=3DArial size=3D2>End MBool.</FONT></DIV>
<DIV><FONT face=3DArial size=3D2></FONT>&nbsp;</DIV>
<DIV><FONT face=3DArial size=3D2>Problems dissapear if we let =
</FONT></DIV>
<DIV>&nbsp;</DIV>
<DIV><FONT face=3DArial size=3D2>
<DIV><FONT face=3DArial size=3D2>Module XBool :I_MBool.</FONT></DIV>
<DIV><FONT face=3DArial size=3D2>End XBool.</FONT></DIV>
<DIV><FONT face=3DArial size=3D2>Module MBool:I_MBool =
:=3DXBool.</FONT></DIV><BR>For=20
full extracted files see&nbsp;<A=20
href=3D"http://www.mathcoq.dva.lv/downloads/Nat.zip">http://www.mathcoq.d=
va.lv/downloads/Nat.zip</A></FONT></DIV>
<DIV><FONT face=3DArial size=3D2></FONT>&nbsp;</DIV>
<DIV><FONT face=3DArial size=3D2>Sorry if all this&nbsp;just =
for&nbsp;lack of my=20
understanding of Ocaml compiling.&nbsp;</FONT></DIV>
<DIV><FONT face=3DArial size=3D2></FONT>&nbsp;</DIV>
<DIV><FONT face=3DArial size=3D2>Best Regards,</FONT></DIV>
<DIV><FONT face=3DArial size=3D2>Jevgenijs.</DIV></FONT>
<DIV><FONT face=3DArial size=3D2></FONT>&nbsp;</DIV>
<DIV><FONT face=3DArial =
size=3D2>**************************************<BR>file=20
indBool.mli<BR>**************************************<BR>module MBool :=20
<BR>&nbsp;I_IndBool.I_MBool</FONT></DIV>
<DIV>&nbsp;</DIV>
<DIV><FONT face=3DArial =
size=3D2>**************************************<BR>file=20
indBool.ml<BR>**************************************</FONT></DIV>
<DIV>&nbsp;</DIV>
<DIV><FONT face=3DArial size=3D2>module MBool =3D <BR>&nbsp;struct =
<BR>&nbsp; type=20
ind_bool =3D<BR>&nbsp;&nbsp;&nbsp; | =
Coq_ind_B_false<BR>&nbsp;&nbsp;&nbsp; |=20
Coq_ind_B_true<BR>......</FONT></DIV>
<DIV>&nbsp;</DIV>
<DIV><FONT face=3DArial size=3D2>end</FONT></DIV>
<DIV>&nbsp;</DIV>
<DIV><FONT face=3DArial =
size=3D2>**************************************<BR>file=20
B_Props.ml<BR>**************************************</FONT></DIV>
<DIV>&nbsp;</DIV>
<DIV><FONT face=3DArial size=3D2>open IndBool<BR>open Setoids<BR>module =
XBoolProps =3D=20
<BR>&nbsp;struct <BR>&nbsp; module BoolCases =3D <BR>&nbsp;&nbsp; struct =

<BR>......<BR>&nbsp; let coq_NOT =3D function<BR>&nbsp;&nbsp;&nbsp; |=20
IndBool.MBool.Coq_ind_B_false -&gt; coq_true<BR>&nbsp;&nbsp;&nbsp; |=20
IndBool.MBool.Coq_ind_B_true -&gt; coq_false<BR>........<BR>&nbsp;&nbsp; =

end<BR>end</FONT></DIV>
<DIV>&nbsp;</DIV>
<DIV><FONT face=3DArial =
size=3D2>**************************************<BR>ocamlc=20
-c&nbsp;&nbsp; B_Props.ml<BR>File "B_Props.ml", line 70, characters=20
10-39:<BR>Unbound constructor=20
IndBool.MBool.Coq_ind_B_false<BR>**************************************</=
FONT></DIV>
<DIV>&nbsp;</DIV><FONT face=3DArial size=3D2>
<DIV><BR>**************************************<BR>file=20
Main_Nat.ml<BR>**************************************<BR>&nbsp; let rec =
coq_rec=20
f a =3D function<BR>&nbsp;&nbsp;&nbsp; | IndBool.MBool.Coq_ind_O -&gt;=20
a<BR>&nbsp;&nbsp;&nbsp; | IndBool.MBool.Coq_ind_S m -&gt; f (coq_rec f a =

m)<BR>&nbsp;end<BR>**************************************<BR>ocamlc=20
-c&nbsp;&nbsp; Main_Nat.ml<BR>File "Main_Nat.ml", line 62, characters=20
6-29:<BR>Unbound constructor IndBool.MBool.Coq_ind_O</DIV>
<DIV>&nbsp;</DIV>
<DIV></FONT>&nbsp;</DIV></BODY></HTML>

------=_NextPart_000_0008_01C4ACB6.43257490--

## Bugzilla Comment ID: 58
Date: 2007-07-04 17:02:32 +0200
From: jevgenijs@dva.lv

a) fixed
b) feature wish, not simple nor priorary
c) ??? 
d) known problem with no simple solution
e) feature wish, impossible for the moment

## Bugzilla Comment ID: 1855
Date: 2007-10-17 15:21:54 +0200
From: @_letouzey

Hi Jevgenijs

The issue d) should now be fixed in Coq trunk (commit 10232). An artificial module is added by extraction around the hidden object in order to always have an unambiguous fully-qualified name for it.

Unfortunately, this fix was done too late for inclusion in 8.1pl2.
But even if this is a big patch, svn diff -c10232 could probably
be applied on 8.1pl2 without too much trouble.

Regards,
Pierre Letouzey

--------------------------
Current Status: 

a) fixed
b) feature wish, not simple nor priorary
c) not enough elements to reproduce (URL with 404), and 
   heavy changes of modular extraction since.
d) fixed
e) feature wish, impossible for the moment

## Bugzilla Comment ID: 2530
Date: 2008-10-21 11:11:50 +0200
From: @_letouzey

Thanks to the work of Elie Soubiran, extraction is now aware of the "with"
annotation in module types. Point e) of this bug report is hence solved.

This leaves only 
b) minor feature wish about empty modules : will not be done, at least 
   not until the next major refactoring of modular extraction.
c) not enough elements to reproduce (URL with 404), and 
   _really_ heavy changes of modular extraction since.

Overall, I consider this bug report as solved, and I'll close it. 

Pierre L.
